hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -239,18 +239,18 @@ object PruneDeadFields {
         val leftDep = left.typ.copy(
           rowType = TStruct(left.typ.rowType.required, left.typ.rowType.fieldNames.flatMap(f =>
             requestedType.rowType.fieldOption(f).map(reqF => f -> reqF.typ)): _*),
-          globalType = requestedType.globalType)
+          globalType = TStruct(left.typ.globalType.required, left.typ.globalType.fieldNames.flatMap(f =>
+            requestedType.globalType.fieldOption(f).map(reqF => f -> reqF.typ)): _*))
         memoizeTableIR(left, leftDep, memo)
-
-        val rightKeyFields = right.typ.key.iterator.flatten.toSet
+        val rightKeyFields = right.typ.key.getOrElse(FastIndexedSeq()).toSet
```

Comment:
`right.typ.keyOrEmpty.toSet` is shorter, and easier to clean up when `key` (soon) becomes required on `TableType`.